aten/src/ATen/native/SpectralOps.cpp

```diff
@@ -17,23 +17,221 @@
 #include <cmath>
 
 namespace at { namespace native {
+static inline Tensor _fft(
+    const Tensor &self, int64_t signal_ndim, bool complex_input,
+    const bool complex_output, bool inverse, IntArrayRef signal_sizes,
+    fft_norm_mode normalization, bool onesided);
+
+namespace {
+
+ScalarType promote_type_fft(ScalarType type, bool require_complex) {
+  if (at::isComplexType(type)) {
+    return type;
+  }
+  // Promote integral to default float type
+  if (!at::isFloatingType(type)) {
+    type = c10::typeMetaToScalarType(c10::get_default_dtype());
+  }
+
+  TORCH_CHECK(type == kFloat || type == kDouble, "Unsupported dtype ", type);
+
+  if (!require_complex) {
+    return type;
+  }
+
+  // Promote to complex
+  switch (type) {
+  case kFloat: return kComplexFloat;
+  case kDouble: return kComplexDouble;
+  }
+  TORCH_INTERNAL_ASSERT(false, "Unhandled dtype");
+}
+
+Tensor promote_tensor_fft(const Tensor& t, bool require_complex=false) {
+  auto cur_type = t.scalar_type();
+  auto new_type = promote_type_fft(cur_type, require_complex);
+  return (cur_type == new_type) ? t : t.to(new_type);
```

Comment:
Currently float tensors (containing complex data) input to fft have shape `(...,2)`, so would `t.to(new_type)` suffice for cases like `promote_type_fft(kFloat, True)`? wouldn't you wanna use `torch.view_as_complex` in such cases?

Reply:
These functions do not treat a `(..., 2)` real tensor as though it were complex. This takes real input and promotes it to a complex tensor with zero imaginary component, regardless of shape.